espsecure.py

```diff
@@ -274,22 +307,27 @@ def _flash_encryption_tweak_key(key, offset, tweak_range):
     """
     if esptool.PYTHON2:
         key = [ord(k) for k in key]
-    else:
-        key = list(key)
-    assert len(key) == 32
+        assert len(key) == 32
 
-    offset_bits = [(offset & (1 << x)) != 0 for x in range(24)]
+        offset_bits = [(offset & (1 << x)) != 0 for x in range(24)]
 
-    for bit in tweak_range:
-        if offset_bits[_FLASH_ENCRYPTION_TWEAK_PATTERN[bit]]:
-            # note that each byte has a backwards bit order, compared
-            # to how it is looked up in the tweak pattern table
-            key[bit // 8] ^= 1 << (7 - (bit % 8))
+        for bit in tweak_range:
+            if offset_bits[_FLASH_ENCRYPTION_TWEAK_PATTERN[bit]]:
+                # note that each byte has a backwards bit order, compared
+                # to how it is looked up in the tweak pattern table
+                key[bit // 8] ^= 1 << (7 - (bit % 8))
+
+        key = b"".join(chr(k) for k in key)
+        return key
 
-    if esptool.PYTHON2:
-        return b"".join(chr(k) for k in key)
     else:
-        return bytes(key)
+        key = int.from_bytes(key, byteorder='big', signed=False)
```

Comment:
This looks like it's probably relatively expensive. Is there a way to take it out of the loop?

Reply:
int.from_byte and int.to_bytes should be rather quick.
bigints are stored as an array of uint32_t and all it has to do is copy some bytes. I chose big-endian was chosen because the data bits were in reverse as well (and this made masking/tweaking pattern order looked more logical as well).

I don't see a faster method for python2 right away. Perhaps suggest users to use python3 with a nice warning or automagically try python3...

```
if __name__ == "__main__":
    if sys.version_info[0] == 2:
        try:
            exit(os.execv('/usr/bin/env', ['env', 'python3', __file__] + sys.argv[1:]))
        except OSError:
            pass
    main()
```

Reply:
well.. big ints are there really, just not the conversion. So it might be easier just to implement these.
I found functions:

```
def to_bytes(n, length, endianess='big'):
    h = '%x' % n
    s = ('0'*(len(h) % 2) + h).zfill(length*2).decode('hex')
    return s if endianess == 'big' else s[::-1]

or a variant

def to_bytes(n, length, endianess='big'):
    h = '%x' % n
    s = codecs.decode(('0'*(len(h) % 2) + h).zfill(length*2), "hex_codec")
    return s if endianess == 'big' else s[::-1]
```
for int.to_bytes. Found some other code here https://gemfury.com/agriconnect/python:cryptography/-/content/utils.py

A slightly modified version here:
```
import binascii

if hasattr(int, "from_bytes"):
    int_from_bytes = int.from_bytes
else:
    def int_from_bytes(data, byteorder, signed=False):
        assert byteorder == 'big'
        assert not signed

        return int(binascii.hexlify(data), 16)


if hasattr(int, "to_bytes"):
    def int_to_bytes(integer, length=None, byteorder='big', signed=False):
        return integer.to_bytes(
            length or (integer.bit_length() + 7) // 8 or 1, byteorder, signed
        )
else:
    def int_to_bytes(integer, length=None, byteorder='big', signed=False):
        assert byteorder == 'big'
        assert not signed

        hex_string = '%x' % integer
        if length is None:
            n = len(hex_string)
        else:
            n = length * 2
        return binascii.unhexlify(hex_string.zfill(n + (n & 1)))
```
which seems to work.

Reply:
Note: You can also speed it up by taking advantage of large multiplications.  It's a minor speed improvement.

```
 @@ -295,6 +295,13 @@ def _flash_encryption_tweak_range_bits(flash_crypt_config=0xF):
     return tweak_range
 
 
+# Forward bit order masks
+mul1        = 0x0000200004000080000004000080001000000200004000080000040000800010
+mul2        = 0x0000000000000000200000000000000010000000000000002000000000000001
+
+mul1_mask   = 0xffffffffffffff801ffffffffffffff00ffffffffffffff81ffffffffffffff0
+mul2_mask   = 0x000000000000007fe00000000000000ff000000000000007e00000000000000f
+
 def _flash_encryption_tweak_key(key, offset, tweak_range):
     """Apply XOR "tweak" values to the key, derived from flash offset
     'offset'. This matches the ESP32 hardware flash encryption.
@@ -322,12 +329,9 @@ def _flash_encryption_tweak_key(key, offset, tweak_range):
 
     else:
         key = int.from_bytes(key, byteorder='big', signed=False)
-        for offset_bit in range(24):
-            if offset & (1 << offset_bit) != 0:
-                key ^= _FLASH_ENCRYPTION_TWEAK_PATTERN_REVERSE[offset_bit] & tweak_range
-
-        key = int.to_bytes(key, length=32, byteorder='big', signed=False)
-        return key
+        addr = offset >> 5
+        key ^= (mul1 * addr) | ((mul2 * addr) & mul2_mask)
+        return int.to_bytes(key, length=32, byteorder='big', signed=False)
 
 
 def generate_flash_encryption_key(args):
```

Additionally, the input_file.tell() call, at least in Linux, invokes a system call per invocation, but the file I/O is buffered in 4k chunks.  You can gain some more time by keeping track of the offset internally.  Take a look at the improvements to the speed I made: https://github.com/tim-nordell-nimbelink/esptool/tree/feature/encryption_speed

Reply:
https://github.com/darkfader/esptool/pull/1/files

I also suspect that it could be a little faster if we used gray code order for encoding, and did a differential xor from address to address generated in gray code order.  This would cause out of order accesses to the files, but might be an overall speed improvement, however, that'd be a bit more complicated.  You could also do chunks of the file in parallel to take advantage of multiple cores in a system, too.  (See https://en.wikipedia.org/wiki/Gray_code).